homeassistant/components/beewi_smartclim/sensor.py

```diff
@@ -0,0 +1,187 @@
+    """Representation of a Sensor."""
+
+    def __init__(self, poller, name, mac, device, unit):
+        """Initialize the sensor."""
+        self._poller = poller
+        self._name = name
+        self._mac = mac
+        self._device = device
+        self._unit = unit
+        self._state = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor. Convert Celsius to Farhenheit if needed."""
+        if self._device == DEVICE_CLASS_TEMPERATURE and self._unit == TEMP_FAHRENHEIT:
+            return (self._state * 1.8) + 32
+        return self._state
+
+    @property
+    def device_class(self):
+        """Device class of this entity."""
+        return self._device
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit
+
+    def update(self):
+        """Fetch new state data from the poller."""
+        self._state = self._poller.get_value(self._device)
+
+
+class BeewiSmartClimPoller:
+    """This class will interact with the sensor and aggregates all data."""
+
+    def __init__(self, mac):
+        """Initialize the Poller."""
+        try:
+            from btlewrap import BluepyBackend
+
+            backend = BluepyBackend
+        except ImportError:
+            from btlewrap import GatttoolBackend
+
+            backend = GatttoolBackend
+
+        self._backend = backend
+        self._mac = mac
+        self._temp = None
+        self._humidity = None
+        self._battery = None
+        self._last_update = None
+
+        _LOGGER.debug("MiTempBtSensorPoller initiated with backend %s", self._backend)
+
+    def get_value(self, device):
+        """
+        Return the value from the cached data.
+
+        In the case it's outdated, it will call the update endpoint to refresh those values.
+        """
+        if (self._last_update is None) or (
+            datetime.now() - timedelta(minutes=3) > self._last_update
+        ):
+            self.update_data()
+        else:
+            _LOGGER.debug("Serving data from cache")
+
+        if device == DEVICE_CLASS_TEMPERATURE:
+            return self._temp
+        if device == DEVICE_CLASS_HUMIDITY:
+            return self._humidity
+        if device == DEVICE_CLASS_BATTERY:
+            return self._battery
+        return None
+
+    def update_data(self):
+        """
+        Get data from device.
+
+        This method reads the handle 0x003f that contains temperature, humidity
+        and battery level.
+        """
+        from btlewrap.base import BluetoothInterface, BluetoothBackendException
+
+        bt_interface = BluetoothInterface(self._backend, "hci0")
+
+        try:
+            with bt_interface.connect(self._mac) as connection:
+                raw = connection.read_handle(0x003F)  # pylint: disable=no-member
```

Comment:
Please extract the device specific interface code, I/O and data parsing, to a standalone library published on pypi.
https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services